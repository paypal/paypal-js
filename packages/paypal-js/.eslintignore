dist/*
scripts/*
v6-docs/*
