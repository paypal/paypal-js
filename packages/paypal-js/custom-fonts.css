@import url("https://fonts.googleapis.com/css2?family=Inter:wght@300;400;500;600;700&family=JetBrains+Mono:ital,wght@0,400;0,500;0,600;1,400;1,500;1,600&display=swap");

:root,
:root[data-theme="light"],
:root[data-theme="dark"] {
    --font-sans: "Inter", -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto,
        "Helvetica Neue", Arial, sans-serif !important;
    --font-mono: "JetBrains Mono", "SF Mono", Monaco, Menlo, Consolas,
        "Liberation Mono", "Courier New", monospace !important;
}

body,
.tsd-page-title,
.tsd-breadcrumb,
.tsd-navigation,
h1,
h2,
h3,
h4,
h5,
h6 {
    font-family:
        "Inter",
        -apple-system,
        BlinkMacSystemFont,
        "Segoe UI",
        Roboto,
        "Helvetica Neue",
        Arial,
        sans-serif !important;
}

code,
pre,
.tsd-signature,
.tsd-code-block {
    font-family: "JetBrains Mono", "SF Mono", Monaco, Menlo, Consolas,
        "Liberation Mono", "Courier New", monospace !important;
}
