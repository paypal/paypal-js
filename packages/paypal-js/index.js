if (process.env.NODE_ENV === "production") {
    module.exports = require("./dist/cjs/paypal-js.min.js");
} else {
    module.exports = require("./dist/cjs/paypal-js.js");
}
