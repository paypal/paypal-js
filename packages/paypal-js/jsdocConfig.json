{
    "source": {
        "include": ["./types/v6/"],
        "includePattern": "\\.(d.ts)$"
    },
    "opts": {
        "destination": "./v6-docs/",
        "recurse": true,
        "readme": "./v6-docs/README.md",
        "template": "../../node_modules/better-docs"
    },
    "plugins": [
        "plugins/markdown",
        "../../node_modules/better-docs/typescript"
    ],
    "tags": {
        "allowUnknownTags": ["optional"]
    },
    "templates": {
        "cleverLinks": false,
        "monospaceLinks": false
    },
    "markdown": {
        "parser": "gfm"
    }
}
