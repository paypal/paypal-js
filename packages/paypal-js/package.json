{
    "name": "@paypal/paypal-js",
    "version": "8.0.2",
    "description": "Loading wrapper and TypeScript types for the PayPal JS SDK",
    "main": "index.js",
    "module": "dist/esm/paypal-js.js",
    "types": "types/index.d.ts",
    "publishConfig": {
        "access": "public"
    },
    "scripts": {
        "build": "rm -rf dist && rollup --config --bundleConfigAsCjs",
        "lint": "eslint .",
        "prerelease": "npm run validate",
        "release": "standard-version --commit-all",
        "postrelease": "git push && git push --follow-tags && npm run build && npm publish",
        "test": "vitest src",
        "test:bundle": "vitest bundle-tests/**",
        "test:e2e": "playwright test",
        "test:e2e:debug": "PWDEBUG=1 playwright test --project=\"chromium\"",
        "test:e2e:start": "node e2e-tests/http-server.js",
        "typecheck": "tsc --noEmit",
        "validate": "npm run check-node-version && npm run format:check && npm run typecheck && npm run build && npm run lint && npm test -- --watch=false --coverage.enabled --coverage.include=src && npm run test:bundle -- --watch=false",
        "openapi": "npm run openapi-orders && npm run openapi-subscriptions",
        "openapi-orders": "openapi-typescript node_modules/@paypal/paypal-rest-api-specifications/openapi/checkout_orders_v2.json -o types/apis/openapi/checkout_orders_v2.d.ts",
        "openapi-subscriptions": "openapi-typescript node_modules/@paypal/paypal-rest-api-specifications/openapi/billing_subscriptions_v1.json -o types/apis/openapi/billing_subscriptions_v1.d.ts"
    },
    "files": [
        "dist",
        "types",
        "!types/**/*.test.ts"
    ],
    "keywords": [
        "paypal",
        "sdk",
        "async",
        "script",
        "loader",
        "types",
        "typescript",
        "typescript-definitions"
    ],
    "license": "Apache-2.0",
    "repository": {
        "type": "git",
        "url": "git://github.com/paypal/paypal-js.git"
    },
    "devDependencies": {
        "@commitlint/cli": "18.4.3",
        "@commitlint/config-conventional": "18.4.3",
        "@paypal/paypal-rest-api-specifications": "^1.0.0",
        "@playwright/test": "^1.40.1",
        "@rollup/plugin-commonjs": "25.0.7",
        "@rollup/plugin-node-resolve": "15.2.3",
        "@rollup/plugin-replace": "5.0.5",
        "@rollup/plugin-terser": "^0.4.4",
        "@rollup/plugin-typescript": "11.1.5",
        "@types/node": "^20.10.5",
        "@types/promise-polyfill": "6.0.6",
        "@typescript-eslint/eslint-plugin": "6.15.0",
        "@typescript-eslint/parser": "6.15.0",
        "@vitest/coverage-v8": "^1.1.0",
        "eslint": "8.56.0",
        "filesize": "10.1.0",
        "husky": "8.0.3",
        "jsdom": "^23.0.1",
        "lint-staged": "15.2.0",
        "openapi-typescript": "^6.7.3",
        "prettier": "3.1.1",
        "rollup": "4.9.1",
        "semver": "7.5.4",
        "standard-version": "9.5.0",
        "tslib": "2.6.2",
        "typescript": "5.3.3",
        "vitest": "1.1.0"
    },
    "dependencies": {
        "promise-polyfill": "^8.3.0"
    }
}
