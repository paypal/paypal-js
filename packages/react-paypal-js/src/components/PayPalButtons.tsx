import React, { useEffect, useRef, useState } from "react";

import { usePayPalScriptReducer } from "../hooks/scriptProviderHooks";
import { getPayPalWindowNamespace, generateErrorMessage } from "../utils";
import { SDK_SETTINGS } from "../constants";

import type { FunctionComponent } from "react";
import type { PayPalButtonsComponent, OnInitActions } from "@paypal/paypal-js";
import type { PayPalButtonsComponentProps } from "../types";

/**
This `<PayPalButtons />` component supports rendering [buttons](https://developer.paypal.com/docs/business/javascript-sdk/javascript-sdk-reference/#buttons) for PayPal, Venmo, and alternative payment methods.
It relies on the `<PayPalScriptProvider />` parent component for managing state related to loading the JS SDK script.
*/
export const PayPalButtons: FunctionComponent<PayPalButtonsComponentProps> = ({
    className = "",
    disabled = false,
    children,
    forceReRender = [],
    ...buttonProps
}: PayPalButtonsComponentProps) => {
    const isDisabledStyle = disabled ? { opacity: 0.38 } : {};
    const classNames = `${className} ${
        disabled ? "paypal-buttons-disabled" : ""
    }`.trim();
    const buttonsContainerRef = useRef<HTMLDivElement>(null);
    const buttons = useRef<PayPalButtonsComponent | null>(null);

    const [{ isResolved, options }] = usePayPalScriptReducer();
    const [initActions, setInitActions] = useState<OnInitActions | null>(null);
    const [isEligible, setIsEligible] = useState(true);
    const [, setErrorState] = useState(null);

    function closeButtonsComponent() {
        if (buttons.current !== null) {
            buttons.current.close().catch(() => {
                // ignore errors when closing the component
            });
        }
    }

    // useEffect hook for rendering the buttons
    useEffect(() => {
        // verify the sdk script has successfully loaded
        if (isResolved === false) {
            return closeButtonsComponent;
        }

        const paypalWindowNamespace = getPayPalWindowNamespace(
            options.dataNamespace
        );

        // verify dependency on window object
        if (
            paypalWindowNamespace === undefined ||
            paypalWindowNamespace.Buttons === undefined
        ) {
            setErrorState(() => {
                throw new Error(
                    generateErrorMessage({
                        reactComponentName: PayPalButtons.displayName as string,
                        sdkComponentKey: "buttons",
                        sdkRequestedComponents: options.components,
                        sdkDataNamespace: options[SDK_SETTINGS.DATA_NAMESPACE],
                    })
                );
            });
            return closeButtonsComponent;
        }

        const decoratedOnInit = (
            data: Record<string, unknown>,
            actions: OnInitActions
        ) => {
            setInitActions(actions);
            if (typeof buttonProps.onInit === "function") {
                buttonProps.onInit(data, actions);
            }
        };

        try {
            buttons.current = paypalWindowNamespace.Buttons({
                ...buttonProps,
                onInit: decoratedOnInit,
            });
        } catch (err) {
            return setErrorState(() => {
                throw new Error(
                    `Failed to render <PayPalButtons /> component. Failed to initialize:  ${err}`
                );
            });
        }

        // only render the button when eligible
        if (buttons.current.isEligible() === false) {
            setIsEligible(false);
            return closeButtonsComponent;
        }

        if (!buttonsContainerRef.current) {
            return closeButtonsComponent;
        }

        buttons.current.render(buttonsContainerRef.current).catch((err) => {
            // component failed to render, possibly because it was closed or destroyed.
            if (
                buttonsContainerRef.current === null ||
                buttonsContainerRef.current.children.length === 0
            ) {
                // paypal buttons container is no longer in the DOM, we can safely ignore the error
                return;
            }
            // paypal buttons container is still in the DOM
            setErrorState(() => {
                throw new Error(
                    `Failed to render <PayPalButtons /> component. ${err}`
                );
            });
        });

        return closeButtonsComponent;
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [isResolved, ...forceReRender, buttonProps.fundingSource]);

    // useEffect hook for managing disabled state
    useEffect(() => {
        if (initActions === null) {
            return;
        }

        if (disabled === true) {
            initActions.disable().catch(() => {
                // ignore errors when disabling the component
            });
        } else {
            initActions.enable().catch(() => {
                // ignore errors when enabling the component
            });
        }
    }, [disabled, initActions]);

    return (
        <>
            {isEligible ? (
                <div
                    ref={buttonsContainerRef}
                    style={isDisabledStyle}
                    className={classNames}
                />
            ) : (
                children
            )}
        </>
    );
};

PayPalButtons.displayName = "PayPalButtons";
