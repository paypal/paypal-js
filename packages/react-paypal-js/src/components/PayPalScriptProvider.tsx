import React, { useEffect, useReducer } from "react";
import { loadScript } from "@paypal/paypal-js";

import {
    getScriptID,
    ScriptContext,
    scriptReducer,
} from "../context/scriptProviderContext";
import { SCRIPT_ID, SDK_SETTINGS, LOAD_SCRIPT_ERROR } from "../constants";
import { SCRIPT_LOADING_STATE, DISPATCH_ACTION } from "../types";

import type { FC } from "react";
import type { ScriptProviderProps } from "../types";

/**
This `<PayPalScriptProvider />` component takes care of loading the JS SDK `<script>`.
It manages state for script loading so children components like `<PayPalButtons />` know when it's safe to use the `window.paypal` global namespace.

Note: You always should use this component as a wrapper for  `PayPalButtons`, `PayPalMarks`, `PayPalMessages` and `BraintreePayPalButtons` components.
 */
export const PayPalScriptProvider: FC<ScriptProviderProps> = ({
    options = { clientId: "test" },
    children,
    deferLoading = false,
}: ScriptProviderProps) => {
    const [state, dispatch] = useReducer(scriptReducer, {
        options: {
            ...options,
            [SDK_SETTINGS.DATA_JS_SDK_LIBRARY]: SDK_SETTINGS.DATA_LIBRARY_VALUE,
            [SDK_SETTINGS.DATA_SDK_INTEGRATION_SOURCE]:
                SDK_SETTINGS.DATA_LIBRARY_VALUE,
            [SCRIPT_ID]: `${getScriptID(options)}`,
        },
        loadingStatus: deferLoading
            ? SCRIPT_LOADING_STATE.INITIAL
            : SCRIPT_LOADING_STATE.PENDING,
    });

    useEffect(() => {
        if (
            deferLoading === false &&
            state.loadingStatus === SCRIPT_LOADING_STATE.INITIAL
        ) {
            return dispatch({
                type: DISPATCH_ACTION.LOADING_STATUS,
                value: SCRIPT_LOADING_STATE.PENDING,
            });
        }

        if (state.loadingStatus !== SCRIPT_LOADING_STATE.PENDING) {
            return;
        }

        let isSubscribed = true;

        loadScript(state.options)
            .then(() => {
                if (isSubscribed) {
                    dispatch({
                        type: DISPATCH_ACTION.LOADING_STATUS,
                        value: SCRIPT_LOADING_STATE.RESOLVED,
                    });
                }
            })
            .catch((err) => {
                console.error(`${LOAD_SCRIPT_ERROR} ${err}`);
                if (isSubscribed) {
                    dispatch({
                        type: DISPATCH_ACTION.LOADING_STATUS,
                        value: {
                            state: SCRIPT_LOADING_STATE.REJECTED,
                            message: String(err),
                        },
                    });
                }
            });
        return () => {
            isSubscribed = false;
        };
    }, [state.options, deferLoading, state.loadingStatus]);

    return (
        <ScriptContext.Provider value={{ ...state, dispatch }}>
            {children}
        </ScriptContext.Provider>
    );
};
