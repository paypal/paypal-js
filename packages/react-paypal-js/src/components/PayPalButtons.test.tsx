import React, { useState } from "react";
import {
    render,
    waitFor,
    screen,
    fireEvent,
    act,
} from "@testing-library/react";
import { ErrorBoundary } from "react-error-boundary";
import { mock } from "jest-mock-extended";
import { loadScript } from "@paypal/paypal-js";

import { PayPalButtons } from "./PayPalButtons";
import { PayPalScriptProvider } from "./PayPalScriptProvider";
import { FUNDING } from "../index";

import type { ReactNode } from "react";
import type {
    PayPalButtonsComponent,
    PayPalButtonsComponentOptions,
    PayPalNamespace,
} from "@paypal/paypal-js";

jest.mock("@paypal/paypal-js", () => ({
    loadScript: jest.fn(),
}));

const onError = jest.fn();
const wrapper = ({ children }: { children: ReactNode }) => (
    <ErrorBoundary fallback={<div>Error</div>} onError={onError}>
        {children}
    </ErrorBoundary>
);

const mockPaypalButtonsComponent = mock<PayPalButtonsComponent>();
mockPaypalButtonsComponent.close.mockResolvedValue();
mockPaypalButtonsComponent.isEligible.mockReturnValue(true);
mockPaypalButtonsComponent.render.mockResolvedValue();

const mockPayPalNamespace = mock<PayPalNamespace>();

mockPayPalNamespace.Buttons = jest
    .fn()
    .mockReturnValue(mockPaypalButtonsComponent);

describe("<PayPalButtons />", () => {
    beforeEach(() => {
        document.body.innerHTML = "";

        window.paypal = mockPayPalNamespace;

        (loadScript as jest.Mock).mockResolvedValue(window.paypal);
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    test("should pass props to window.paypal.Buttons()", async () => {
        render(
            <PayPalScriptProvider options={{ clientId: "test" }}>
                <PayPalButtons
                    fundingSource={FUNDING.CREDIT}
                    style={{ layout: "horizontal" }}
                />
            </PayPalScriptProvider>
        );

        await waitFor(() =>
            expect(window.paypal?.Buttons).toHaveBeenCalledWith({
                style: { layout: "horizontal" },
                fundingSource: FUNDING.CREDIT,
                onInit: expect.any(Function),
            })
        );
    });

    test("should use className prop and add to div container", async () => {
        render(
            <PayPalScriptProvider options={{ clientId: "test" }}>
                <PayPalButtons className="custom-class-name" />
            </PayPalScriptProvider>
        );

        await waitFor(() =>
            expect(document.querySelector("div.custom-class-name")).toBeTruthy()
        );
    });

    test("should disable the Buttons with disabled=true", async () => {
        const onInitCallbackMock = jest.fn();

        render(
            <PayPalScriptProvider options={{ clientId: "test" }}>
                <PayPalButtons
                    className="custom-class-name"
                    disabled={true}
                    onInit={onInitCallbackMock}
                />
            </PayPalScriptProvider>
        );

        await waitFor(() => {
            expect(
                document.querySelector("div.paypal-buttons-disabled")
            ).toBeTruthy();
        });
        expect(window.paypal?.Buttons).toHaveBeenCalled();

        const onInitActions = {
            enable: jest.fn().mockResolvedValue(true),
            disable: jest.fn().mockResolvedValue(true),
        };

        act(() =>
            (window.paypal?.Buttons as jest.Mock).mock.calls[0][0].onInit(
                {},
                onInitActions
            )
        );

        await waitFor(() => {
            expect(onInitCallbackMock).toHaveBeenCalled();
        });
        expect(onInitActions.disable).toHaveBeenCalled();
    });

    test("should enable the Buttons when disabled=true", async () => {
        const { container, rerender } = render(
            <PayPalScriptProvider options={{ clientId: "test" }}>
                <PayPalButtons
                    className="custom-class-name"
                    disabled={true}
                    onInit={jest.fn()}
                />
            </PayPalScriptProvider>
        );

        await waitFor(() => {
            expect(
                container
                    .querySelector("div.custom-class-name")
                    ?.classList.contains("paypal-buttons-disabled")
            ).toBeTruthy();
        });
        const onInitActions = {
            enable: jest.fn().mockResolvedValue(true),
            disable: jest.fn().mockResolvedValue(true),
        };

        act(() =>
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            (window.paypal?.Buttons as jest.Mock).mock.calls[0][0]?.onInit(
                {},
                onInitActions
            )
        );
        rerender(
            <PayPalScriptProvider options={{ clientId: "test" }}>
                <PayPalButtons
                    className="custom-class-name"
                    disabled={false}
                    onInit={onInitActions.enable}
                />
            </PayPalScriptProvider>
        );
        await waitFor(() => {
            expect(onInitActions.enable).toBeCalled();
        });
        expect(
            container
                ?.querySelector("div.custom-class-name")
                ?.classList.contains("paypal-buttons-disabled")
        ).toBeFalsy();
    });

    test("should re-render Buttons when any item from props.forceReRender changes", async () => {
        function ButtonWrapper({ initialAmount }: { initialAmount: string }) {
            const [amount, setAmount] = useState(initialAmount);
            const [currency, setCurrency] = useState("USD");
            return (
                <>
                    <button onClick={() => setAmount(amount + 1)}>
                        Update Amount
                    </button>
                    <button onClick={() => setCurrency("EUR")}>
                        Update Currency
                    </button>
                    <PayPalButtons forceReRender={[amount, currency]} />
                </>
            );
        }

        render(
            <PayPalScriptProvider options={{ clientId: "test" }}>
                <ButtonWrapper initialAmount="1" />
            </PayPalScriptProvider>
        );

        await waitFor(() =>
            expect(window.paypal?.Buttons).toHaveBeenCalledTimes(1)
        );

        fireEvent.click(screen.getByText("Update Amount"));

        // confirm re-render when the forceReRender value changes for first item
        await waitFor(() =>
            expect(window.paypal?.Buttons).toHaveBeenCalledTimes(2)
        );

        fireEvent.click(screen.getByText("Update Currency"));

        // confirm re-render when the forceReRender value changes for second item
        await waitFor(() =>
            expect(window.paypal?.Buttons).toHaveBeenCalledTimes(3)
        );
    });
    window.paypal?.Buttons as jest.Mock;
    test("should not re-render Buttons from side-effect in props.createOrder function", async () => {
        function ButtonWrapper({ initialOrderID }: { initialOrderID: string }) {
            const [orderID, setOrderID] = useState(initialOrderID);
            return (
                <>
                    <div data-testid="orderID">{orderID}</div>
                    <PayPalButtons
                        createOrder={() => {
                            setOrderID("2");
                            return Promise.resolve("2");
                        }}
                    />
                </>
            );
        }

        render(
            <PayPalScriptProvider options={{ clientId: "test" }}>
                <ButtonWrapper initialOrderID="1" />
            </PayPalScriptProvider>
        );

        await waitFor(() =>
            expect(window.paypal?.Buttons).toHaveBeenCalledTimes(1)
        );

        expect(screen.getByTestId("orderID").innerHTML).toBe("1");

        await act(() =>
            // call createOrder() to trigger a state change
            (window.paypal?.Buttons as jest.Mock).mock.calls[0][0].createOrder()
        );

        await waitFor(() =>
            expect(screen.getByTestId("orderID").innerHTML).toBe("2")
        );

        // confirm that the Buttons were NOT reset by a side-effect in createOrder()
        expect(window.paypal?.Buttons).toHaveBeenCalledTimes(1);
    });

    test("should render custom content when ineligible", async () => {
        window.paypal = {
            Buttons: jest.fn(() => ({
                close: jest.fn().mockResolvedValue(true),
                isEligible: jest.fn().mockReturnValue(false),
            })),
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
        } as any;

        render(
            <PayPalScriptProvider options={{ clientId: "test" }}>
                <PayPalButtons className="test-button" />
            </PayPalScriptProvider>
        );

        // defaults to rendering null when ineligible
        await waitFor(() => {
            expect(window.paypal?.Buttons).toBeCalled();
        });
        expect(
            (window.paypal?.Buttons as jest.Mock).mock.results[0].value
                .isEligible
        ).toBeCalled();
        expect(
            (window.paypal?.Buttons as jest.Mock).mock.results[0].value
                .isEligible.mock.results[0].value
        ).toBeFalsy();
    });

    test("should throw an error when no components are passed to the PayPalScriptProvider", async () => {
        const spyConsoleError = jest
            .spyOn(console, "error")
            .mockImplementation();
        // reset the paypal namespace to trigger the error
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        window.paypal = {} as any;

        render(
            <PayPalScriptProvider options={{ clientId: "test" }}>
                <PayPalButtons />
            </PayPalScriptProvider>,
            { wrapper }
        );

        await waitFor(() => expect(onError).toHaveBeenCalled());
        expect(onError.mock.calls[0][0].message).toMatchSnapshot();
        spyConsoleError.mockRestore();
    });

    test("should throw an error when the 'buttons' component is missing from the components list passed to the PayPalScriptProvider", async () => {
        const spyConsoleError = jest
            .spyOn(console, "error")
            .mockImplementation();
        // reset the paypal namespace to trigger the error
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        window.paypal = {} as any;

        render(
            <PayPalScriptProvider
                options={{
                    clientId: "test",
                    components: "marks,messages",
                }}
            >
                <PayPalButtons />
            </PayPalScriptProvider>,
            { wrapper }
        );

        await waitFor(() => expect(onError).toHaveBeenCalled());
        expect(onError.mock.calls[0][0].message).toMatchSnapshot();
        spyConsoleError.mockRestore();
    });

    test("should catch and throw unexpected zoid render errors", async () => {
        const spyConsoleError = jest
            .spyOn(console, "error")
            .mockImplementation();
        window.paypal = {
            Buttons() {
                return {
                    close: jest.fn().mockResolvedValue({}),
                    isEligible: jest.fn().mockReturnValue(true),
                    render: jest.fn((element: string | HTMLElement) => {
                        // simulate adding markup for paypal button
                        if (typeof element !== "string") {
                            element.append(document.createElement("div"));
                        }
                        return Promise.reject("Unknown error");
                    }),
                };
            },
            version: "",
        };

        render(
            <PayPalScriptProvider options={{ clientId: "test" }}>
                <PayPalButtons />
            </PayPalScriptProvider>,
            { wrapper }
        );

        await waitFor(() => expect(onError).toHaveBeenCalled());
        expect(onError.mock.calls[0][0].message).toMatchSnapshot();
        spyConsoleError.mockRestore();
    });

    test("should throw an error during initialization when style prop is invalid", async () => {
        const spyConsoleError = jest
            .spyOn(console, "error")
            .mockImplementation();
        window.paypal = {
            Buttons(options?: PayPalButtonsComponentOptions | undefined) {
                if (
                    options?.style?.color === "gold" &&
                    options?.fundingSource === "venmo"
                ) {
                    throw new Error(
                        "Unexpected style.color for venmo button: gold, expected blue, silver, black, white"
                    );
                }
                return mockPaypalButtonsComponent;
            },
            version: "",
        };

        render(
            <PayPalScriptProvider
                options={{
                    clientId: "test",
                    components: "marks,messages",
                }}
            >
                <PayPalButtons
                    style={{ color: "gold" }}
                    fundingSource={FUNDING.VENMO}
                />
            </PayPalScriptProvider>,
            { wrapper }
        );

        await waitFor(() => expect(onError).toBeCalled());
        expect(onError.mock.calls[0][0]).toEqual(
            expect.objectContaining({
                message:
                    "Failed to render <PayPalButtons /> component. Failed to initialize:  Error: Unexpected style.color for venmo button: gold, expected blue, silver, black, white",
            })
        );
        spyConsoleError.mockRestore();
    });

    test("should safely ignore error on render process when paypal buttons container is no longer in the DOM ", async () => {
        const spyConsoleError = jest
            .spyOn(console, "error")
            .mockImplementation();
        const mockRender = jest
            .fn()
            .mockRejectedValue(new Error("Unknown error"));
        window.paypal = {
            Buttons() {
                return {
                    close: jest.fn().mockResolvedValue({}),
                    isEligible: jest.fn().mockReturnValue(true),
                    render: mockRender,
                };
            },
            version: "",
        };

        const { container } = render(
            <PayPalScriptProvider options={{ clientId: "test" }}>
                <PayPalButtons className="test-children" />
            </PayPalScriptProvider>
        );

        await waitFor(() => expect(mockRender).toBeCalled());
        expect(
            container.querySelector("div.test-children")?.hasChildNodes()
        ).toBeFalsy();
        spyConsoleError.mockRestore();
    });
});
