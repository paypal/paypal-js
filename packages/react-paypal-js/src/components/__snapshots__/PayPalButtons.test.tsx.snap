// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PayPalButtons /> should catch and throw unexpected zoid render errors 1`] = `"Failed to render <PayPalButtons /> component. Unknown error"`;

exports[`<PayPalButtons /> should throw an error when no components are passed to the PayPalScriptProvider 1`] = `
"Unable to render <PayPalButtons /> because window.paypal.Buttons is undefined.
To fix the issue, add 'buttons' to the list of components passed to the parent PayPalScriptProvider:
\`<PayPalScriptProvider options={{ components: 'buttons'}}>\`."
`;

exports[`<PayPalButtons /> should throw an error when the 'buttons' component is missing from the components list passed to the PayPalScriptProvider 1`] = `
"Unable to render <PayPalButtons /> because window.paypal.Buttons is undefined.
To fix the issue, add 'buttons' to the list of components passed to the parent PayPalScriptProvider:
\`<PayPalScriptProvider options={{ components: 'marks,messages,buttons'}}>\`."
`;
